### PR TITLE
C++17 fallback.

### LINKED ISSUE #971: Isolate the bug with simplified lambda expressions
Some of the code in sample.h and observe.h uses lambda expressions that are C++20 (so cannot be backported) but also cause failures in the CI. The failure are related to the async test with g++ 12 and debug builds.

### PR DESCRIPTION
**Pull request by schweitzpgi**
These changes modify the CUDAQ headers such that they can be compiled by a C++ compiler in C++17 mode.

Simplify the diffs in observe.h, sample.h

Fix #971. Do perfect forwarding into the lambda captures instead of passing references. This is required for asynchronous calls.

Augment the tests in NVQPP to use the -std=c++17 command line option. Not all of these will execute the resulting c++17 executable in order to keep the test time down.

### CODE REVIEW

**Review by boschmitt — COMMENTED** [review]
Thanks for all this work Eric.

File: tools/cudaq-quake/cudaq-quake.cpp [hunk 1]
```diff
@@ -111,6 +111,11 @@ static cl::list<std::string>
 static cl::list<std::string>
     extraClangArgs("Xcudaq", cl::desc("Extra options to pass to clang++"));
 
+static cl::opt<std::string> stdCpp(
+    "std",
+    cl::desc("Specify the C++ standard (c++17, c++20). The default is c++20."),
+    cl::init("c++20"));
+
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply tools/cudaq-quake/cudaq-quake.cpp:118]: Good idea. I'm not sure this is the final list however. C++11 and C++14 may or may not need to be on the list.

File: runtime/cudaq/qis/qvector.h [hunk 2]
```diff
@@ -53,25 +54,47 @@ class qvector {
...
+  /// @returns the `[start, start+size)` qudits as a non-owning `qview`.
   qview<Levels> slice(std::size_t start, std::size_t size) {
+#if CUDAQ_USE_STD20
     return std::span(qudits).subspan(start, size);
+#else
+    typename std::vector<value_type>::const_iterator first =
+        qudits.begin() + start;
+    typename std::vector<value_type>::const_iterator last =
+        qudits.begin() + start + size;
+    return {qudits(first, last)};
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply runtime/cudaq/qis/qvector.h:96]: It's not a call operator. It's a constructor.

But this was some code from a batch of changes that I later revised. I'll clean this up, thanks.

**Comment by schweitzpgi** [comment]
> Thanks for all this work Eric.
> 
> Edit: We can minimize the use of `#if CUDAQ_USE_STD20` if the c++20 and c++17 implementations of `qview` share a couple of constructors, namely:
> 
> ```c++
>   template <typename Iterator>
>   qview(Iterator &&otherQudits, std::size_t otherQusize)
>       : qudits(&*otherQudits), qusize(otherQusize) {}
>   template <typename R>
>   qview(R &&other)
>       : qudits(&*other.begin()),
>         qusize(std::distance(other.begin(), other.end())) {}
> ```
> 
> On the c++20 version, it would use these to construct the span within the class; Then, for example, we would be able to implement `back(...)` in `qreg` without the need to specialize on the c++ version:
> 
> ```c++
>   qspan<dyn, Levels> back(std::size_t count) {
>     return {qudits.end() - count, count};
>   }
> ```
> 
> This also applies for `qspan`.

So, a couple of points come to mind here.

- I didn't want to change the original C++ '20 code or expand its interfaces in any way. I made a few exceptions to this in the algorithms files that removed C++ '20 code and replaced it with C++ '11 code where they should (hopefully) be non-functional changes.

- In `qview` and `qspan`, we had to reimplement these from the ground up since they rely of `std::span` which is only valid in C++ '20 and later. In order to thread that needle a couple of constructors were added to the interfaces. Those should not be part of the C++ '20 interface and are only there as glue so `qvector`, etc. can build and return spans (of subranges) in its member functions.

I think that the use of any C++ standard prior to 20 should be discouraged/deprecated, since that was a foundational assumption for the CUDA Quantum project.

**Review by boschmitt — APPROVED** [review]

**Pull request merged**